docs/proposals/role_decomposition.md

```diff
@@ -0,0 +1,322 @@
+# Scaffolding for decomposing large roles
+
+## Why?
+
+Currently we have roles that are very large and encompass a lot of different
+components. This makes for a lot of logic required within the role, can
+create complex conditionals, and increases the learning curve for the role.
+
+## How?
+
+Creating a guide on how to approach breaking up a large role into smaller,
+component based, roles. Also describe how to develop new roles, to avoid creating
+large roles.
+
+## Proposal
+
+Create a new guide or append to the current contributing guide a process for
+identifying large roles that can be split up, and how to compose smaller roles
+going forward.
+
+### Large roles
+
+A role should be considered for decomposition if it:
+
+1) Configures/installs more than one product.
+1) Can configure multiple variations of the same product that can live
+side by side.
+1) Has different entry points for upgrading and installing a product
```

Comment:
This can be tricky. E.g. we have a role for master installation. We have tasks that should go inside a new master upgrade role. Lot of the tasks are common (configuration files and system services logic, restart handlers, installation/updating deps(rpms, images), etc.). Given the current ansible has still some issues wrt. resource sharing (e.g can not include handlers from different roles properly), it is non-trivial task to break install/upgrade role into two install and upgrade roles without increasing redundancy.